Net/NetClient.cpp

```diff
@@ -0,0 +1,209 @@
+		HandlePacket(packetReceiveBuffer, packetSize);
+		break;
+	case PacketType::FRAGMENT:
+		const auto [fragmentCount, fragmentIndex] = ReadFragmentPacketSubHeader(&packetReceiveBuffer[BASE_HEADER_SIZE]);
+		ReceivedFragmentPacketData* fragmentPacketData = &receivedFragmentPackets[sequence % MAX_FRAGMENT_PACKET_BUFFER_SIZE];
+
+		if ((int)fragmentCount > MAX_FRAGMENTS || fragmentCount == 0 || fragmentIndex >= fragmentCount)
+		{
+			// TODO: Log?
+			// TODO: Metrics?
+			return;
+		}
+
+		if (fragmentPacketData->sequence != sequence)
+			InitNewFragmentPacketData(sequence, fragmentCount);
+
+		if (!fragmentPacketData->fragmentReceived[fragmentIndex])
+		{
+			memcpy(
+				&fragmentPacketData->packetData[(uint64_t)fragmentIndex * MAX_FRAGMENT_PAYLOAD_SIZE],
+				&packetReceiveBuffer[BASE_HEADER_SIZE + FRAGMENT_SUB_HEADER_SIZE],
+				packetSize - BASE_HEADER_SIZE - FRAGMENT_SUB_HEADER_SIZE);
+			fragmentPacketData->fragmentReceived[fragmentIndex] = true;
+			++fragmentPacketData->fragmentReceivedCount;
+
+			if (fragmentIndex == fragmentCount - 1)
+			{
+				const int lastFragmentSize = packetSize - BASE_HEADER_SIZE - FRAGMENT_SUB_HEADER_SIZE;
+				fragmentPacketData->totalSize = (fragmentCount - 1) * MAX_FRAGMENT_PAYLOAD_SIZE + lastFragmentSize;
+			}
+		}
+
+		if (fragmentPacketData->fragmentReceivedCount == fragmentPacketData->fragmentCount)
+			HandlePacket(fragmentPacketData->packetData, fragmentPacketData->totalSize);
+
+		break;
+	}
+}
+
+void NetClient::WriteBasePacketHeader(uint8_t* packetBuffer, const BasePacketHeader* header)
+{
+	assert(packetBuffer);
+	assert(header);
+
+	const uint16_t netSequence = htons(header->sequence);
+	packetBuffer[0] = (netSequence >> 8) & 0xFF;
+	packetBuffer[1] = netSequence & 0xFF;
+
+	packetBuffer[2] = (uint8_t)header->packetType;
+}
+
+void NetClient::WriteFragmentPacketHeader(uint8_t* packetBuffer, const FragmentPacketSubHeader* header)
+{
+	assert(packetBuffer);
+	assert(header);
+
+	packetBuffer[0] = header->fragmentCount;
+	packetBuffer[1] = header->fragmentIndex;
+}
+
+NetClient::BasePacketHeader NetClient::ReadBasePacketHeader(const uint8_t* packetBuffer)
+{
+	BasePacketHeader baseHeader;
+	const auto packetType = (PacketType)packetBuffer[2];
+	const uint16_t hostSequence = (uint16_t)(packetBuffer[0] << 8) | packetBuffer[1];
+
+	baseHeader.sequence = ntohs(hostSequence);
+	baseHeader.packetType = packetType;
+
+	return baseHeader;
+}
+
+NetClient::FragmentPacketSubHeader NetClient::ReadFragmentPacketSubHeader(const uint8_t* packetBuffer)
+{
+	FragmentPacketSubHeader fragmentHeader;
+
+	fragmentHeader.fragmentCount = packetBuffer[0];
+	fragmentHeader.fragmentIndex = packetBuffer[1];
+
+	return fragmentHeader;
+}
+
+void NetClient::HandlePacket(uint8_t* packet, int packetSize)
+{
+	// TEST
+	for (uint32_t i = 0; (uint64_t)i < packetSize / sizeof(int); ++i)
+		if (i != ((uint32_t*)packet)[i])
+		{
+			DL_NET_LOG_ERROR("Error in packet data");
+			return;
+		}
+
+	DL_NET_LOG_ERROR("Success!");
```

Comment:
Remove this test code. It is unclear how the packet handling will be done at the moment so it could be delayed until the next layer has been implemented.